### PR TITLE
Improve the SDSServer features

### PR DESCRIPTION
**Pull request by mangchiandjjoe**
- Implemented DeregisterUdsPath
- Added data protection with Mutex
- Return the same version number for the same X.509 key/cert

### CODE REVIEW

File: security/pkg/workload/server.go [hunk 1]
```diff
@@ -190,16 +180,34 @@ func (s *SDSServer) RegisterUdsPath(udsPath string) error {
...
 	return nil
 }
 
-// TODO: add methods to unregister the deleted UDS paths and the listeners.
+// DeregisterUdsPath closes and removes the grpcServer instance serving UDS
+func (s *SDSServer) DeregisterUdsPath(udsPath string) error {
+	s.udsServerMapGuard.Lock()
+	defer s.udsServerMapGuard.Unlock()
+
+	if udsServer, ok := s.udsServerMap[udsPath]; ok {
```
Comment (wattli) [inline security/pkg/workload/server.go]: I feel the logic can be made more readable :)

if !ok {
   return error...
}
everything else
return nil

File: security/pkg/workload/server.go [hunk 2]
```diff
@@ -86,15 +84,16 @@ func (s *SDSServer) SetServiceIdentityCert(content []byte) error {
...
 }
 
 // GetTLSCertificate generates the X.509 key/cert for the workload identity
 // derived from udsPath, which is where the FetchSecrets grpc request is
 // received.
-// SecretServer implementations could have diffent implementation
-func (s *SDSServer) GetTLSCertificate(udsPath string) (*auth.TlsCertificate, error) {
+// SecretServer implementations could have different implementation
+func (s *SDSServer) GetTLSCertificate() (*auth.TlsCertificate, error) {
 	s.certificateChainGuard.RLock()
```
Comment (wattli) [inline security/pkg/workload/server.go]: Do we also need to provide root cert?

File: security/pkg/workload/server.go [hunk 3]
```diff
@@ -61,12 +49,21 @@ type SDSServer struct {
...
+	// Each UDS path identifies the identity for which the workload will
+	// request X.509 key/cert from this server. This path should only be
+	// accessible by such workload.
+	udsServerMap map[string]*grpc.Server
+
+	// Mutex for udsServerMap
+	udsServerMapGuard sync.Mutex
+
+	// current certificate chain and private key version number
+	version string
```
Comment (myidpt) [inline security/pkg/workload/server.go]: What is this version used for?

File: security/pkg/workload/server.go [hunk 4]
```diff
@@ -61,12 +49,21 @@ type SDSServer struct {
 
 	// Read/Write mutex for privateKey
 	privateKeyGuard sync.RWMutex
```
Comment (myidpt) [inline security/pkg/workload/server.go]: Can we use KeyCertBundle for better protection of the key and certs? https://github.com/istio/istio/blob/master/security/pkg/pki/util/keycertbundle.go

File: security/pkg/workload/server.go (hunk 3 above)
Comment (mangchiandjjoe) [reply security/pkg/workload/server.go]: DiscoveryResponse response requires version_info to check certificates were updated.
When the version increased, subscription module of the Envoy calls onConfigUpdate function.

https://github.com/envoyproxy/data-plane-api/blob/78b7960a01bd1ad5248cb8de230ddb04ccf6dfd4/envoy/api/v2/discovery.proto#L60

File: security/pkg/workload/server.go (hunk 2 above)
Comment (mangchiandjjoe) [reply security/pkg/workload/server.go]: We don't need the root cert for now. Secret.TlsCertificate only requires certificate chain and private key.

https://github.com/envoyproxy/data-plane-api/blob/master/envoy/api/v2/auth/cert.proto#L74

File: security/pkg/workload/server.go (hunk 1 above)
Comment (mangchiandjjoe) [reply security/pkg/workload/server.go]: Done

File: security/pkg/workload/server.go (hunk 4 above)
Comment (mangchiandjjoe) [reply security/pkg/workload/server.go]: I will do that in another PR

**Comment by mangchiandjjoe** [comment]
@wattli @myidpt PTAL

File: security/pkg/workload/server.go [hunk 5]
```diff
@@ -190,16 +180,35 @@ func (s *SDSServer) RegisterUdsPath(udsPath string) error {
...
+
 	// grpcServer.Serve() is a blocking call, so run it in a goroutine.
 	go func() {
-		log.Infof("Starting GRPC server on UDS path", udsPath)
-
-		err := s.grpcServer.Serve(listener)
+		log.Infof("Starting GRPC server on UDS path: %s", udsPath)
+		err := udsServer.Serve(listener)
 		// grpcServer.Serve() always returns a non-nil error.
 		log.Warnf("GRPC server returns an error: %v", err)
```
Comment (myidpt) [inline security/pkg/workload/server.go]: Add a TODO here to pass the error back to the caller, so that correct operations can be conducted.

**Pull request merged**